File: apps/mobile/modules/player-webview/ios/PlayerwebviewModule.swift
import ExpoModulesCore

public class PlayerwebviewModule: Module {
  public func definition() -> ModuleDefinition {
    Name("PlayerWebview")

    // ── View manager ──
    View(PlayerWebView.self) {
      Events("onLoadingStart", "onLoadingFinish", "onHttpError", "onMessage")

      Prop("source") { (view: PlayerWebView, value: [String: Any]) in
        if let uri = value["uri"] as? String {
          view.sourceUri = uri
        }
      }

      Prop("userAgent") { (view: PlayerWebView, value: String) in
        view.userAgent = value
      }

      Prop("forceLoadUrl") { (view: PlayerWebView, value: String) in
        view.forceLoadUrl = value
      }

      Prop("injectedJavaScriptBeforeContentLoaded") { (view: PlayerWebView, value: String) in
        view.injectedScript = value
      }

      Prop("allowsFullscreenVideo") { (_: PlayerWebView, _: Bool) in
        // Fullscreen is always allowed via WKUIDelegate
      }

      Prop("setSupportMultipleWindows") { (view: PlayerWebView, value: Bool) in
        view.supportMultipleWindows = value
      }

      Prop("referrer") { (view: PlayerWebView, value: String) in
        view.referrer = value
      }

      Prop("javaScriptCanOpenWindowsAutomatically") { (view: PlayerWebView, value: Bool) in
        view.javaScriptCanOpenWindowsAutomatically = value
      }

      // ── Imperative control props (setNativeProps) ──
      Prop("forceReload") { (view: PlayerWebView, _: Double) in
        view.reloadView()
      }

      Prop("forceStop") { (view: PlayerWebView, _: Double) in
        view.stop()
      }

      Prop("injectedJavaScript_") { (view: PlayerWebView, value: String) in
        view.injectJavaScript(value)
      }

      // ── Lifecycle ──
      OnViewDidUpdateProps { _ in
        // Props are handled individually via their didSet observers
      }
    }

    // ── Clear all WebView storage (cookies, cache, localStorage) ──
    AsyncFunction("clearAllState") {
      await clearWebViewState()
    }
  }

  private func clearWebViewState() async {
    // WKWebsiteDataStore removes all data including cookies, cache, localStorage, IndexedDB
    let dataTypes = WKWebsiteDataStore.allWebsiteDataTypes()
    let since = Date.distantPast

    await WKWebsiteDataStore.default().removeData(
      ofTypes: dataTypes,
      modifiedSince: since
    )
  }
}

File: apps/mobile/modules/player-webview/android/src/main/java/expo/modules/playerwebview/PlayerwebviewModule.kt
package expo.modules.playerwebview

import android.webkit.CookieManager
import android.webkit.WebStorage
import android.webkit.WebView
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class PlayerWebviewModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("PlayerWebview")

    // Window-Overlay WebView — attaches the real WebView to the Activity window,
    // bypassing Fabric's compositing pipeline entirely.
    // File: PlayerWebViewOverlayView.kt
    View(PlayerWebViewOverlayView::class) {
      Events(
        "onLoadingStart",
        "onLoadingFinish",
        "onHttpError",
        "onMessage",
        "onRenderProcessGone",
        "onAuditData"
      )

      Prop("injectedJavaScriptBeforeContentLoaded") { view: PlayerWebViewOverlayView, value: String ->
        view.injectedScript = value
      }

      Prop("source") { view: PlayerWebViewOverlayView, value: Map<String, Any> ->
        val uri = value["uri"] as? String ?: return@Prop
        view.sourceUri = uri
      }

      Prop("userAgent") { view: PlayerWebViewOverlayView, value: String ->
        view.userAgent = value
      }

      Prop("injectedJavaScriptAfterLoad") { view: PlayerWebViewOverlayView, value: String ->
        view.injectedJavaScriptAfterLoad = value
      }

      Prop("allowsFullscreenVideo") { _: PlayerWebViewOverlayView, _: Boolean -> }

      Prop("setSupportMultipleWindows") { view: PlayerWebViewOverlayView, value: Boolean ->
        view.supportMultipleWindows = value
      }

      Prop("referrer") { view: PlayerWebViewOverlayView, value: String ->
        view.referrer = value
      }

      Prop("javaScriptCanOpenWindowsAutomatically") { view: PlayerWebViewOverlayView, value: Boolean ->
        view.javaScriptCanOpenWindowsAutomatically = value
      }

      Prop("forceLoadUrl") { view: PlayerWebViewOverlayView, value: String ->
        view.forceLoadUrl = value
      }

      Prop("forceReload") { view: PlayerWebViewOverlayView, _: Double ->
        view.reload()
      }

      Prop("forceStop") { view: PlayerWebViewOverlayView, _: Double ->
        view.stop()
      }

      Prop("injectedJavaScript_") { view: PlayerWebViewOverlayView, value: String ->
        view.injectedJavaScript_ = value
      }

      Prop("auditMode") { view: PlayerWebViewOverlayView, value: Boolean ->
        view.auditMode = value
      }
    }

    AsyncFunction("clearAllState") {
      clearWebViewState()
    }
  }

  private fun clearWebViewState() {
    val context = appContext.reactContext ?: return
    CookieManager.getInstance().removeAllCookies(null)
    CookieManager.getInstance().flush()
    WebStorage.getInstance().deleteAllData()
    WebView(context).apply {
      clearCache(true)
      clearFormData()
      clearHistory()
      clearSslPreferences()
      destroy()
    }
  }
}
